boa/docs/boa.8
.TH BOA 8 "Jan 22 2000" "Version 0.94"
.SH NAME
.B boa \- a single\-tasking high performance http server
.SH SYNOPSIS
.B boa
.RB [ -c 
.IR server_root ] 
.RB [ -r 
.IR chroot ] 
.RB [ -d ]
.SH DESCRIPTION
Boa is a single-tasking HTTP server. That means that unlike traditional web
servers, it does not fork for each incoming connection, nor does it fork many
copies of itself to handle multiple connections. It internally multiplexes all
of the ongoing HTTP connections, and forks only for CGI programs (which must be
separate processes.) Preliminary tests show Boa is more than twice as fast as
Apache.
.PP
The primary design goals of Boa are speed and security. Security, in the sense
of "can't be subverted by a malicious user", not "fine grained access control
and encrypted communications". Boa is not intended as a feature-packed server;
if you want one of those, check out WN from John Franks. Modifications to Boa
that improve its speed, security, robustness, and portability, are eagerly
sought. Other features may be added if they can be achieved without hurting the
primary goals. 
.SH OPTIONS
.IP \fB-d\fR 
instruct Boa not to fork itself (non-daemonize).

.IP "\fB-c \fIserver_root\fR"
choose a server root overriding the default SERVER_ROOT #define in 
.I defines.h

The server root must hold your local copy of the configuration file

.IP "\fB-r \fIchroot\fR"
instruct Boa where to chdir and chroot to.  The chdir/chroot
is done before the configuration file is read, or any log
files are opened.

.SH FILES
.TP
\fBboa.conf\fR \- the sole configuration file for Boa.  
The directives in this file are defined in the 
.B DIRECTIVES
section.

.TP
\fBmime.types\fR \- the 
MimeTypes <filename>
defines what Content-Type Boa will send in an HTTP/1.0
or better transaction.
.SH DIRECTIVES

The Boa configuration file is parsed with a lex/yacc or flex/bison generated
parser.  If it reports an error, the line number will be provided; it should
be easy to spot.  The syntax of each of these rules is very simple, and they
can occur in any order.  Where possible, these directives mimic those of NCSA
httpd 1.3; We saw no reason to introduce gratuitous differences.
.PP
Note: the "ServerRoot" is not in this configuration file.  It can be compiled
into the server (see 
.I defines.h
) or specified on the command line with the 
.B -c 
option.

The following directives are contained in the 
.I boa.conf
file, and most, but not all, are required.
.TP
Port <integer>
This is the port that Boa runs on.  The default port for http servers is 80. 
If it is less than 1024, the server must be started as root.
.TP
User <user name or UID>
The name or UID the server should run as.  For Boa to attempt this, the
server must be started as root.
.TP
Group <group name or GID>
The group name or GID the server should run as.  For Boa to attempt this,
the server must be started as root.
.TP
ServerAdmin <email address>
The email address where server problems should be sent.
Note: this is not currently used. 
.TP
PidFile <filename>
Where to put the pid of the process.
Comment out to write no pid file.
Note: Because Boa drops privileges at startup, and the
pid file is written by the UID/GID before doing so, Boa
does not attempt removal of the pid file.
.TP
ErrorLog <filename>
The location of the error log file.  If this does not start with
/, it is considered relative to the server root.
Set to /dev/null if you don't want errors logged. 
.TP
AccessLog <filename>
The location of the access log file.  If this does not start with /, it is
considered relative to the server root.
Comment out or set to /dev/null (less effective) to disable access logging.
.TP
VerboseCGILogs
This is a logical switch and does not take any parameters.
Comment out to disable.
.TP
CGILog <filename>
The location of the CGI error log file.  If this does not start with /, it 
is considered relative to the server root. If specified, this is the file 
that the stderr of CGIs is tied to, *instead* of to the ErrorLog.
.TP
CGIumask <umask>
The CGIumask is set immediately before execution of the CGI.
The default value is 027. The number must be interpretable
unambiguously by the C function strtol. No base is specified,
so one may use a hexadecimal, decimal, or octal number if 
it is prefixed accordingly.
.TP
ServerName <server_name>
The name of this server that should be sent back to
clients if different than that returned by gethostname.
.TP
VirtualHost
This is a logical switch and does not take any parameters.
Comment out to disable.
Given DocumentRoot /var/www, requests on interface 'A' or IP 'IP-A'
become /var/www/IP-A.
Example: http://localhost/ becomes /var/www/127.0.0.1 
.TP
VHostRoot <directory>
The root location for all virtually hosted data
Comment out to disable.
Incompatible with 'Virtualhost' and 'DocumentRoot'!!
Given VHostRoot /var/www, requests to host foo.bar.com,
where foo.bar.com is ip a.b.c.d,
become /var/www/a.b.c.d/foo.bar.com 
Hostnames are "cleaned", and must conform to the rules
specified in rfc1034, which are be summarized here:

Hostnames must start with a letter, end with a letter or digit, 
and have as interior characters only letters, digits, and hyphen.
Hostnames must not exceed 63 characters in length.
.TP
DefaultVHost <hostname>
Define this in order to have a default hostname when the client does not
specify one, if using VirtualHostName. If not specified, the word
"default" will be used for compatibility with older clients.
.TP
DocumentRoot <directory>
The root directory of the HTML documents. If this does not start with
/, it is considered relative to the server root.
.TP
UserDir <directory>
The name of the directory which is appended onto a user's home directory if a
~user request is received.
.TP
DirectoryIndex <filename>
Name of the file to use as a pre-written HTML directory index.  Please  make
and use these files.  On the fly creation of directory indexes can be slow.
.TP
DirectoryMaker <directory>
Name of the program used to generate on-the-fly directory listings.
The program must take one or two command-line arguments, the first
being the directory to index (absolute), and the second, which is optional,
contains what Boa would have the "title" of the document be.
Comment out if you don't want on the fly directory listings.
If this does not start with
/, it is considered relative to the server root.
.TP
KeepAliveMax <integer>
Number of KeepAlive requests to allow per connection.  Comment out, or set
to 0 to disable keepalive processing. 
.TP
KeepAliveTimeout <integer>
Number of seconds to wait before keepalive connections time out.
.TP
MimeTypes <file>
The location of the 
.I mime.types
file.  If this does not start with /, it is considered relative to 
the server root. Set to /dev/null if you do not want to load a mime types 
file. Do *not* comment out (better use AddType!)
.TP
DefaultType <mime type>
MIME type used if the file extension is unknown, or there is no file extension.
.TP
AddType <mime type> <extension> [extension...]
Associates a MIME type with an extension or extensions.
.TP
Redirect, Alias, and ScriptAlias <path1> <path2>
Redirect, Alias, and ScriptAlias all have the same semantics \-\- they
match the beginning of a request and take appropriate action.  Use
Redirect for other servers, Alias for the same server, and ScriptAlias to
enable directories for script execution.

Redirect allows you to tell clients about documents which used to exist
in your server's namespace, but do not anymore.  This allows you tell
the clients where to look for the relocated document.

Alias aliases one path to another.  Of course, symbolic links in the
file system work fine too.          

ScriptAlias maps a virtual path to a directory for serving scripts.
.PP
Please see the included 
.I boa.conf 
for defaults and examples.
.SH HISTORY
Like the Linux kernel, even numbered versions are "stable", and odd numbered
versions are "unstable", or rather, "development".
Versions 0.91 and 0.91beta of Boa were released by Paul Phillips <paulp@go2net.com>
.PP
Version 0.92 was released by Larry Doolittle on 
December 12, 1996.
.PP
Version 0.93 was the development version of 0.94. 
.PP
Version 0.94 was released 22 Jan 2000.
.SH BUGS
There are probably bugs, but we are not aware of any at this time.
.SH AUTHOR
Boa was created by Paul Phillips <paulp@go2net.com>. 
It is now being maintained and enhanced by Larry Doolittle
<ldoolitt@boa.org> and Jon Nelson <jnelson@boa.org>.
.PP
Linux is the development platform at the moment, other
OS's are known to work. If you'd like to
contribute to this effort, contact Larry or Jon via e-mail. 
.SH LICENSE
This program is distributed under the GNU General Public License, as noted in
each source file.


boa/CHANGES
** Changes from 0.94.13 to 0.94.14
 * SUBSTANTIALLY UPDATE AUTOCONF SYSTEM
   - all of the autoconf stuff is now located in the top level
     directory, and creates the appropriate Makefiles for building
     the system and documentation.
 * add [optional, Linux-only] check for sendfile system call
 * add new state, IOSHUFFLE, which utilizes sendfile (or
   emulates it using the request's buffer, otherwise)
 * make the default socket size 32K.  The client_stream_size stays constant
   at 8K, and the new 'buffer' size is 4K
   Note also that the new code uses the "default" socket buffer size
   it obtains with the first accept()ed socket.
 * use setjmp, instead of sigsetjmp (we dont mess with the signal mask
   so why bother!  it's one less syscall)
 * explicitly set pending_requests to 0 in select.c when sigterm_flag has
   been set, and also when the server socket is *not* set despite checking
   for it.
 * make default behavior be to leave stderr alone, but tie it
   to cgilog otherwise
 * add initial vhost_root support
 * remove support for normalize_path
 * use --disable-sendfile to disable sendfile support
   sendfile support for files > 100K is default, now.
   It's significantly faster and easier than the alternative
 * add "Host" header support to CGI environment
 * umask ~0770 before exec
 * backport hash.c from 0.95 and use fnv1a hash (see CREDITS)
 * don't set stderr close-on-exec
 * make some minor for very useful optimizations in the read/write
   loop for CGIs
 * make some adjustments for Solaris and other platforms
 * add yyerror function definition to boa_lexer.l
 * tie stdout to the access_log, unless there is no access_log,
   in which case tie it to /dev/null
 * use sensible defaults for umask (077) and (027 for CGI)
 * add and document new parameters (CGILog and CGIumask)
 * vast improvements to the cgi-test.cgi program, by Jon Nelson
   and Landon Curt Noll.
 * next 3 items by Don Mahurin (patches modified somewhat):
 * pidfile patch
 * default mime_types patch
 * NCSA environment environment variables wrapped in
   #ifdef USE_NCSA_CGI_ENV
 * fix some escaping issues with the directory indexer
   (Ulf Harnhammar)
 * poll or select available at compile time
 * as of 0.94.14rc2 through 0.94.14rc3:
 * remove support for lex/yacc (backported from 0.95)
 * add conditional support for http/1.1
 * as of 0.94.14rc4:
 * replace many #defines with enumerations
 * update specfile (thanks to Supreet Sethi via SourceForge)
 * support optional config file name argument
 * tentatively support Ranges
 * as of 0.94.14rc5 through 0.94.14rc7:
 * dummy install target and some cleanups for *BSD
 * update index_dir.c to be *really* safe WRT html and http escaping*
   * Note to self - please make that code less ugly
 * basic access (Allow/Deny) control
 * fix check-for-broken-setuid check in boa.c -- root *is* a legal user
 * as of 0.94.14rc9:
 * fix reversed argument passing in add_mime_type
 * disable-gunzip support
 * capitalization changes, etc.. to configure.in
 * use slightly newer AC_DEFINE 3-argument style in configure.in
 * add new member of struct, bytes_written, and use it to more accurately
   report the number of bytes actually written to the socket.
 * When sendfile(2) reports ECONNRESET treat it like EPIPE, which is to
   say silently shut it down and don't be noisy about logging it.
 * wrap sa_family_t typedef in
   #ifdef DONT_HAVE_SA_FAMILY_T
   to deal with non-POSIX (1g) systems (Cygwin?)
 * add req->bytes_written member variable, and use it instead of filepos
 * apply Peter Korsgaard's "configure --help" patch
 * apply Peter Korsgaard's "configure --disable-gunzip" patch
 * use newer 3-argument AC_DEFINE
 * update config.sub and config.guess
 * clean up and update configure.in somewhat
 * don't close stderr in terminal signal handlers
 * free range pool and server_name in (final) sigterm handler
 * refactor select and poll loops
 * as of 0.94.14rc10:
 * change to using PF_ prefix instead of AF_ prefix (PF_ is POSIX?)
 * fix copying too-much-memory (read-side, write-side was OK) in
   ascii_sockaddr.  Also optimize slightly (note, may actually be a wee
   slower due to strlen check -- is there a way around this?)
 * fix a few shadow variable problems and improve select and poll loops
 * check for and use madvise (may or may not help)
 * update depends (the cause of SIGSEGs when ./configure was re-run to
   switch from select to poll, *without* running make clean)
 * always remove select.o and poll.o (and access.o) because these are
   the usual conditional files.
 * if we aren't using IPv6, define BOA_NI_MAXHOST to 20. 1025 is
   far too huge for a single IPv4 IP address in dotted-quad notation.
 * move access_init to earlier in the config reading
 * allow a MAX_FILE_MMAP value of 0 to mean "always mmap"
 * as of 0.94.14rc11:
 * fix 'HEAD' requests (bug introduced in rc6)
 * as of 0.94.14rc12:
 * *huge* patch to try to eliminate shadow variables, use unsigned ints
   instead of signed ints, and "const" where appropriate.
 * check for and use __func__, a C99 construct that is used
   in the DIE and WARN macros to also describe the name of the current
   function.
 * fix a very minor IPv6 issue, and include netdb.h in compat.h so that
   NI_MAXHOST is defined for IPv6
 * when we can't mmap a file, fall back to IOSHUFFLE.  If we couldn't
   mmap the file due to an error in mmap or madvise, report the error,
   otherwise it is safe to assume we simply ran out of hash table space.
   This was reported by Davide Pagnin
 * The following 2 changes borrowed from Hydra (which is itself
   based on Boa):
   * Some optimizations in HTTP header parsing.
   * Added DefaultCharset configuration directive. If given, the default
     character set will be appended to the Content-Type for all
     'text' mime types.
 * update boa.texi with some missing configuration directives
 * as of 0.94.14rc13:
 * force select.c, poll.c, and access.c to always be included in

boa/src/boa.c
/*
 *  Boa, an http server
 *  Copyright (C) 1995 Paul Phillips <paulp@go2net.com>
 *  Some changes Copyright (C) 1996 Charles F. Randall <crandall@goldsys.com>
 *  Copyright (C) 1996-1999 Larry Doolittle <ldoolitt@boa.org>
 *  Copyright (C) 1996-2005 Jon Nelson <jnelson@boa.org>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 1, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

/* $Id: boa.c,v 1.99.2.26 2005/02/22 14:11:29 jnelson Exp $*/

#include "boa.h"

/* globals */
int backlog = SO_MAXCONN;
time_t start_time;

int debug_level = 0;
int sighup_flag = 0;            /* 1 => signal has happened, needs attention */
int sigchld_flag = 0;           /* 1 => signal has happened, needs attention */
int sigalrm_flag = 0;           /* 1 => signal has happened, needs attention */
int sigterm_flag = 0;           /* lame duck mode */
time_t current_time;
int pending_requests = 0;

extern const char *config_file_name;

/* static to boa.c */
static void usage(const char *programname);
static void parse_commandline(int argc, char *argv[]);
static void fixup_server_root(void);
static int create_server_socket(void);
static void drop_privs(void);

static int sock_opt = 1;
static int do_fork = 1;

int main(int argc, char *argv[])
{
    int server_s;               /* boa socket */
    pid_t pid;

    /* set umask to u+rw, u-x, go-rwx */
    /* according to the man page, umask always succeeds */
    umask(077);

    /* but first, update timestamp, because log_error_time uses it */
    (void) time(&current_time);

    /* set timezone right away */
    tzset();

    {
        int devnullfd = -1;
        devnullfd = open("/dev/null", 0);

        /* make STDIN point to /dev/null */
        if (devnullfd == -1) {
            DIE("can't open /dev/null");
        }

        if (dup2(devnullfd, STDIN_FILENO) == -1) {
            DIE("can't dup2 /dev/null to STDIN_FILENO");
        }

        (void) close(devnullfd);
    }

    parse_commandline(argc, argv);
    fixup_server_root();
    read_config_files();
    create_common_env();
    open_logs();
    server_s = create_server_socket();
    init_signals();
    build_needs_escape();

    /* background ourself */
    if (do_fork) {
        pid = fork();
    } else {
        pid = getpid();
    }

    switch (pid) {
    case -1:
        /* error */
        perror("fork/getpid");
        exit(EXIT_FAILURE);
    case 0:
        /* child, success */
        break;
    default:
        /* parent, success */
        if (pid_file != NULL) {
            FILE *PID_FILE = fopen(pid_file, "w");
            if (PID_FILE != NULL) {
                fprintf(PID_FILE, "%d", pid);
                fclose(PID_FILE);
            } else {
                perror("fopen pid file");
            }
        }

        if (do_fork)
            exit(EXIT_SUCCESS);
        break;
    }

    drop_privs();
    /* main loop */
    timestamp();

    status.requests = 0;
    status.errors = 0;

    start_time = current_time;
    loop(server_s);
    return 0;
}

static void usage(const char *programname)
{
    fprintf(stderr,
            "Usage: %s [-c serverroot] [-d] [-f configfile] [-r chroot]%s\n",
            programname,
#ifndef DISABLE_DEBUG
            " [-l debug_level]"
#else
            ""
#endif
        );
#ifndef DISABLE_DEBUG
    print_debug_usage();
#endif
    exit(EXIT_FAILURE);

}

static void parse_commandline(int argc, char *argv[])
{
    int c;                      /* command line arg */

    while ((c = getopt(argc, argv, "c:dl:f:r:")) != -1) {
        switch (c) {
        case 'c':
            if (server_root)
                free(server_root);
            server_root = strdup(optarg);
            if (!server_root) {
                perror("strdup (for server_root)");
                exit(EXIT_FAILURE);
            }
            break;
        case 'd':
            do_fork = 0;
            break;
        case 'f':
            config_file_name = optarg;
            break;
        case 'r':
            if (chdir(optarg) == -1) {
                log_error_time();
                perror("chdir (to chroot)");
                exit(EXIT_FAILURE);
            }
            if (chroot(optarg) == -1) {
                log_error_time();
                perror("chroot");
                exit(EXIT_FAILURE);
            }
            if (chdir("/") == -1) {
                log_error_time();
                perror("chdir (after chroot)");
                exit(EXIT_FAILURE);
            }
            break;
#ifndef DISABLE_DEBUG
        case 'l':
            parse_debug(optarg);
            break;
#endif
        default:
            usage(argv[0]);
            exit(EXIT_FAILURE);
        }
    }
}

static int create_server_socket(void)
{
    int server_s;

    server_s = socket(SERVER_PF, SOCK_STREAM, IPPROTO_TCP);
    if (server_s == -1) {
        DIE("unable to create socket");
    }

    /* server socket is nonblocking */
    if (set_nonblock_fd(server_s) == -1) {
        DIE("fcntl: unable to set server socket to nonblocking");
    }

    /* close server socket on exec so CGIs can't write to it */
    if (fcntl(server_s, F_SETFD, 1) == -1) {
        DIE("can't set close-on-exec on server socket!");
    }

    /* reuse socket addr */
    if ((setsockopt(server_s, SOL_SOCKET, SO_REUSEADDR, (void *) &sock_opt,
                    sizeof (sock_opt))) == -1) {
        DIE("setsockopt");
    }

    /* Internet family-specific code encapsulated in bind_server()  */
    if (bind_server(server_s, server_ip, server_port) == -1) {
        DIE("unable to bind");
    }

    /* listen: large number just in case your kernel is nicely tweaked */
    if (listen(server_s, backlog) == -1) {
        DIE("unable to listen");
    }
    return server_s;
}

static void drop_privs(void)
{
    /* give away our privs if we can */
    if (getuid() == 0) {
        struct passwd *passwdbuf;
        passwdbuf = getpwuid(server_uid);
        if (passwdbuf == NULL) {
            DIE("getpwuid");
        }
        if (initgroups(passwdbuf->pw_name, passwdbuf->pw_gid) == -1) {
            DIE("initgroups");
        }
        if (setgid(server_gid) == -1) {
            DIE("setgid");
        }
        if (setuid(server_uid) == -1) {
            DIE("setuid");
        }
        /* test for failed-but-return-was-successful setuid
         * http://www.securityportal.com/list-archive/bugtraq/2000/Jun/0101.html
         */
        if (server_uid != 0 && setuid(0) != -1) {
            DIE("icky Linux kernel bug!");
        }
    } else {
        if (server_gid || server_uid) {
            log_error_time();
            fprintf(stderr, "Warning: "
                    "Not running as root: no attempt to change"
                    " to uid %u gid %u\n", server_uid, server_gid);
        }
        server_gid = getgid();
        server_uid = getuid();
    }
}

/*
 * Name: fixup_server_root
 *
 * Description: Makes sure the server root is valid.
 *
 */

static void fixup_server_root()
{
    if (!server_root) {
#ifdef SERVER_ROOT
        server_root = strdup(SERVER_ROOT);
        if (!server_root) {
            perror("strdup (SERVER_ROOT)");
            exit(EXIT_FAILURE);
        }
#else
        fputs("boa: don't know where server root is.  Please #define "
              "SERVER_ROOT in boa.h\n"
              "and recompile, or use the -c command line option to "
              "specify it.\n", stderr);
        exit(EXIT_FAILURE);
#endif
    }

    if (chdir(server_root) == -1) {
        fprintf(stderr, "Could not chdir to \"%s\": aborting\n",
                server_root);
        exit(EXIT_FAILURE);
    }
}
